examples/webview2/src/main.rs

```diff
@@ -0,0 +1,622 @@
+            let method = match status {
+                0 => "resolve",
+                _ => "reject",
+            };
+            let js = format!(
+                r#"
+                window._rpc[{}].{}({});
+                window._rpc[{}] = undefined;"#,
+                id, method, result, id
+            );
+
+            webview.eval(&js).expect("eval return script");
+        })
+    }
+
+    fn set_window_webview(hwnd: HWND, webview: Box<WebView>) {
+        unsafe {
+            WindowsAndMessaging::SetWindowLongPtrA(
+                hwnd,
+                WINDOW_LONG_PTR_INDEX::GWLP_USERDATA,
+                Box::into_raw(webview) as _,
+            );
+        }
+    }
+
+    fn get_window_webview(hwnd: HWND) -> Option<Box<WebView>> {
+        unsafe {
+            let data =
+                WindowsAndMessaging::GetWindowLongPtrA(hwnd, WINDOW_LONG_PTR_INDEX::GWLP_USERDATA);
+
+            match data {
+                0 => None,
+                _ => {
+                    let webview_ptr = data as *mut WebView;
+                    let raw = Box::from_raw(webview_ptr);
+                    let webview = raw.clone();
+                    mem::forget(raw);
+
+                    Some(webview)
+                }
+            }
+        }
+    }
+}
+
+fn set_process_dpi_awareness() -> Result<()> {
+    let code = unsafe {
+        HiDpi::SetProcessDpiAwareness(PROCESS_DPI_AWARENESS::PROCESS_PER_MONITOR_DPI_AWARE)
+    };
+
+    if code.is_err() {
+        Err(Error::fast_error(code))
+    } else {
+        Ok(())
+    }
+}
+
+extern "system" fn window_proc(hwnd: HWND, msg: u32, w_param: WPARAM, l_param: LPARAM) -> LRESULT {
+    let webview = match WebView::get_window_webview(hwnd) {
+        Some(webview) => webview,
+        None => return unsafe { WindowsAndMessaging::DefWindowProcA(hwnd, msg, w_param, l_param) },
+    };
+
+    let frame = webview
+        .frame
+        .as_ref()
+        .expect("should only be called for owned windows");
+
+    match msg {
+        WindowsAndMessaging::WM_SIZE => {
+            let size = get_window_size(hwnd);
+            webview
+                .controller
+                .SetBounds(Rect {
+                    X: 0f32,
+                    Y: 0f32,
+                    Width: size.cx as f32,
+                    Height: size.cy as f32,
+                })
+                .expect("call SetBounds");
+            *frame.size.lock().expect("lock size") = size;
+            LRESULT(0)
+        }
+
+        WindowsAndMessaging::WM_CLOSE => {
+            unsafe {
+                WindowsAndMessaging::DestroyWindow(hwnd);
+            }
+            LRESULT(0)
+        }
+
+        WindowsAndMessaging::WM_DESTROY => {
+            webview.terminate().expect("window is gone");
+            LRESULT(0)
+        }
+
+        _ => unsafe { WindowsAndMessaging::DefWindowProcA(hwnd, msg, w_param, l_param) },
+    }
+}
+
+fn wait_with_pump<T>(rx: mpsc::Receiver<T>) -> Result<T> {
```

Comment:
While, in general I'd like to see more comments so that those unfamiliar with WebView2 can use this example to learn more, I think this function in particular could use more explanatory prose.

Reply:
Also, is this a common enough need for it to potentially make sense inside of the windows crate itself?

Reply:
> Also, is this a common enough need for it to potentially make sense inside of the windows crate itself?

Maybe? It's not uncommon for Win32 programs to embed shorter-lived message loops while waiting for an async operation or a synchronization primitive (e.g. using [MsgWaitForMultipleObjects](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-msgwaitformultipleobjectsex)], so the UI remains responsive. The WebView2 threading model runs everything on the UI thread, including callbacks which it triggers with `PostMessage`, and I'm using this here because it's waiting for async operations in WebView2 to finish before I've started the main message loop in `run`.

The trouble with using an embedded message loop like this from within the main message loop is that you can easily trigger reentrancy bugs on the main thread (you're already in a call to `DispatchMessage` from the main thread, your embedded loop dispatches another message, and a message handler which triggered the embedded loop gets called again, changes state, embedded loop finishes, world explodes when the handler that called it resumes). So if this were included in the `windows`  crate, I'd probably want to at least mark it as unsafe and include warnings about not using it in the main loop.

If it's not there, the main loop would be a useful thing to abstract in the `windows` crate, though.

A generic window_proc would be interesting too. Most C/C++ window procs are implemented with `switch`/`case`, the `match` idiom in Rust is a lot less fragile. To get good static analysis, it might even make sense to implement it as a macro rather than a dispatch table mapping messages to `fn`, so the caller implements it with `match` arms inline. Or maybe it could just be another `fn` to implement the `match` expression.

Reply:
I'm working on adding the same level of async support that C++/WinRT provides to the Windows crate. At that point you would not need these pumping loops at all and can simply use a Rust async function. That already works to some degree in Rust except that Rust async functions are cold-start whereas we need hot-start for WinRT. So for now, I'm fine with this function but I'll hopefully get rid of it soon as its just a stopgap to get the sample working.